source/common/upstream/health_discovery_service.cc

```diff
@@ -158,8 +200,15 @@ void HdsDelegate::processMessage(
         ClusterConnectionBufferLimitBytes);
 
     // Add endpoints to cluster
-    auto* endpoints = cluster_config.mutable_load_assignment()->add_endpoints();
     for (const auto& locality_endpoints : cluster_health_check.locality_endpoints()) {
+      // add endpoint group by locality to config
+      auto* endpoints = cluster_config.mutable_load_assignment()->add_endpoints();
+      // if this group contains locality information, save it.
+      if (locality_endpoints.has_locality()) {
+        endpoints->mutable_locality()->MergeFrom(locality_endpoints.locality());
```

Comment:
Why isn't this a required field?

Reply:
This is a great question, I'm not entirely sure why its not required in the specifier. I suppose it could be because it was not necessary information to complete the health checks, but it is just for easier identification on the side of the consumer of the HDS results, when the result is returned. Some of the old tests additionally do not hold locality information in the specifier and do not add the locality information. I suppose for the use case when you really do only need a flat list of endpoints, i.e. one cluster with one locality, it may not be particularly useful, but it is certainly not a detriment.

I'm not sure if it is your intent with this note to maybe add a required annotation to the locality field which it looks like would be a breaking change, since it is part of the API already in Envoy before this PR. I was thinking if we wanted to do this then we could add it to v4alpha, but I couldn't find any sort of validation upgrade annotation in the docs or through my code searching for v4+ similar to the oneof_promotion or deprecated annotations but for validation. Do you know of such a thing? It seems the only one I could find makes the requirement go into effect immediately.

Reply:
Yeah, this validation upgrade annotation doesn't exist. I think for now let's leave it as is.